src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -16,124 +16,204 @@
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.HashMap;
+import java.util.Collection;
 import java.util.stream.Collectors;
 
+
 @RestController
 @CrossOrigin(exposedHeaders = "errors, content-type")
 @RequestMapping("api/import")
 public class ImportCSV {
 
+    public enum ClinicPetAction {
+        ADD, DELETE
+    }
+
+    public static class PetRowResult {
+        public PetRowResult(Pet pet, ClinicPetAction action) {
+            this.pet = pet;
+            this.action = action;
+        }
+
+        public final Pet pet;
+        public final ClinicPetAction action;
+    }
+
+
     @Autowired
     private ClinicService clinicService;
 
+    /**
+     * Parses the whole csv file String. The format is each line has to be separated by ';',
+     * whereas each line has five elements in the following structure:
+     * <p>
+     * PET_NAME;PET_BIRTHDAY;PET_TYPE;PET_OWNER;PET_CLINIC_ACTION_COMMAND\n
+     * <p>
+     * Each line converts to one addition or deletion of an existing pet.
+     *
+     * @param csv The string which should contain the loaded csv file.
+     * @return A matrix in form of an ArrayList<ArrayList<String>>
+     */
+    static public ArrayList<ArrayList<String>> parseCsvFile(String csv) {
+        ArrayList<ArrayList<String>> dataElements = new ArrayList<>();
+        ArrayList<String> currentRow = new ArrayList<>();
+        dataElements.add(currentRow);
+        int oldIndex = 0;
+        for (int i = 0; i < csv.length(); ++i) {
+            final boolean isEndOfLine = csv.charAt(i) == '\n';
+            final boolean isEndOfFile = i + 1 == csv.length();
+            final boolean isSeparator = csv.charAt(i) == ';';
+            if (isSeparator || isEndOfLine) {
+                currentRow.add(csv.substring(oldIndex, i));
+                oldIndex = i + 1;
+            } else if (isEndOfFile) {
+                currentRow.add(csv.substring(oldIndex, i + 1));
+            }
+            if (isEndOfFile || isEndOfLine) {
+                dataElements.add(currentRow);
+                currentRow = new ArrayList<>();
+            }
+        }
+        return dataElements;
+    }
+
+    /**
+     * Converts a single data row from the csv file to a PetRowResult, which contains a Pet and a ClinicPetAction.
+     *
+     * @param csvRow     The csv row containing the pet_name, pet_birthday, pet_type, pet_owner and pet clinic action
+     * @param petTypeMap A mapping of all preexisting pet types to the given pet_type instance in the csvRow
+     * @param petOwners  A collection of all existing pet owners
+     * @return A PetRowResult containing a new pet and the action to perform in the clinic service
+     * @throws Exception if the csvRow does not contain the correct information
+     */
+    static public PetRowResult parsePetRow(ArrayList<String> csvRow, HashMap<String, PetType> petTypeMap, Collection<Owner> petOwners) throws Exception {
```

Comment:
this method maybe looks a bit long and handles too much so it might decrease the readability a bit and make testing a bit more difficult...maybe some parts can be extract to new methods such as parsing each field